devices/bmp180/bmp180.go

```diff
@@ -0,0 +1,258 @@
+	if err := d.readReg(regOutMSB, pressureBuf[:]); err != nil {
+		return err
+	}
+
+	UP := (int32(pressureBuf[0])<<16 + int32(pressureBuf[1])<<8 | int32(pressureBuf[2])) >> (8 - d.os)
+
+	pressure := d.c.compensatePressure(UP, int32(UT), d.os)
+
+	env.Temperature = devices.Celsius(temp * 100)
+	env.Pressure = devices.KPascal(pressure)
+	return nil
+}
+
+// Halt is a noop for the BMP180.
+func (d *Dev) Halt() error {
+	return nil
+}
+
+// Opts is optional options to pass to the constructor.
+type Opts struct {
+	Pressure Oversampling
+}
+
+const (
+	bmp180Address = 0x77 // the address of a BMP180 is fixed at 0x77
+	chipID        = 0x55 // contents of the ID register, always 0x55
+)
+
+// New returns an object that communicates over I²C to BMP180 environmental
+// sensor.
+func New(b i2c.Bus, opts *Opts) (*Dev, error) {
+	d := &Dev{
+		d:  &i2c.Dev{Bus: b, Addr: bmp180Address},
+		os: No,
+	}
+
+	if opts != nil {
+		d.os = opts.Pressure
+	}
+
+	if err := d.makeDev(opts); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+var defaults = Opts{
+	Pressure: O4x,
+}
+
+// registers
+const (
+	regChipID           = 0xD0 // register contains the chip id
+	regCalibrationStart = 0xAA // first calibration register address
+	regSoftReset        = 0xE0 // soft reset register
+	regCtrlMeas         = 0xF4 // control measurement
+	regOutMSB           = 0xF6 // MSB of measurement
+	regOutLSB           = 0xF7 // LSB of measurement
+	regOutXLSB          = 0xF8 // extended LSB of measurement
+)
+
+const softResetValue = 0xB6
+
+func (d *Dev) makeDev(opts *Opts) error {
+	if opts == nil {
+		opts = &defaults
+	}
+	var id [1]byte
+	if err := d.readReg(regChipID, id[:]); err != nil {
+		return err
+	}
+
+	if id[0] != chipID {
+		return fmt.Errorf("bmp180: unexpected chip id 0x%x; is this a BMP180?", id[0])
+	}
+
+	// issue soft reset to initialize device
+	if err := d.d.Tx([]byte{regSoftReset, softResetValue}, nil); err != nil {
+		return err
+	}
+
+	// wait for restart
+	time.Sleep(10 * time.Millisecond)
+
+	// read calibration data from internal EEPROM, 11 registers with two bytes each
+	var cal [22]byte
+	if err := d.readReg(regCalibrationStart, cal[:]); err != nil {
+		return err
+	}
+
+	// consistency check as suggested by the datasheet (page 13)
+	for i := 0; i < len(cal)/2; i++ {
+		val := uint16(cal[2*i])<<8 | uint16(cal[2*i+1])
+		if val == 0 || val == 0xffff {
+			return fmt.Errorf("calibration byte %v has invalid value 0x%x", i, val)
+		}
+	}
+
+	d.c.AC1 = int16(cal[0])<<8 | int16(cal[1])
+	d.c.AC2 = int16(cal[2])<<8 | int16(cal[3])
+	d.c.AC3 = int16(cal[4])<<8 | int16(cal[5])
+	d.c.AC4 = uint16(cal[6])<<8 | uint16(cal[7])
+	d.c.AC5 = uint16(cal[8])<<8 | uint16(cal[9])
+	d.c.AC6 = uint16(cal[10])<<8 | uint16(cal[11])
+	d.c.B1 = int16(cal[12])<<8 | int16(cal[13])
+	d.c.B2 = int16(cal[14])<<8 | int16(cal[15])
+	d.c.MB = int16(cal[16])<<8 | int16(cal[17])
+	d.c.MC = int16(cal[18])<<8 | int16(cal[19])
+	d.c.MD = int16(cal[20])<<8 | int16(cal[21])
+
```

Comment:
To me the `convert` idea is pretty slick. It could actually be a method on `calibration` that takes the byte array as arg and returns a `calibration` struct, mhh, could actually be newCalibration. You're creating a binary.BigEndian reader for the entire struct, not for each 2-byte element. Or am I not getting it right?